pkg/aggregateeventrecorder/aggregateeventrecorder_test.go

```diff
@@ -0,0 +1,216 @@
+package aggregateeventrecorder_test
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lassie/pkg/aggregateeventrecorder"
+	"github.com/filecoin-project/lassie/pkg/events"
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/context"
+)
+
+var testCid1 = mustCid("bafybeihrqe2hmfauph5yfbd6ucv7njqpiy4tvbewlvhzjl4bhnyiu6h7pm")
+
+func TestAggregateEventRecorder(t *testing.T) {
+	var req datamodel.Node
+	var path string
+	receivedChan := make(chan bool, 1)
+	authHeaderValue := "applesauce"
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		req, err = ipld.DecodeStreaming(r.Body, dagjson.Decode)
+		require.NoError(t, err)
+		path = r.URL.Path
+		require.Equal(t, "Basic applesauce", r.Header.Get("Authorization"))
+		receivedChan <- true
+	}))
+	defer ts.Close()
+
+	tests := []struct {
+		name string
+		exec func(t *testing.T, ctx context.Context, subscriber types.RetrievalEventSubscriber, id types.RetrievalID, etime, ptime time.Time, spid peer.ID)
+	}{
+		{
+			name: "Retrieval Success",
+			exec: func(t *testing.T, ctx context.Context, subscriber types.RetrievalEventSubscriber, id types.RetrievalID, etime, ptime time.Time, spid peer.ID) {
+				fetchPhaseStartTime := ptime.Add(-10 * time.Second)
+
+				subscriber(events.Started(id, fetchPhaseStartTime, types.FetchPhase, types.RetrievalCandidate{RootCid: testCid1}))
+				subscriber(events.FirstByte(id, ptime, types.RetrievalCandidate{RootCid: testCid1}))
+				subscriber(events.Success(id, ptime, types.NewRetrievalCandidate(spid, testCid1), uint64(2020), 3030, 4*time.Second, big.Zero()))
+				subscriber(events.Finished(id, fetchPhaseStartTime, types.RetrievalCandidate{RootCid: testCid1}))
+
+				select {
+				case <-ctx.Done():
+					t.Fatal(ctx.Err())
+				case <-receivedChan:
+				}
+
+				require.Equal(t, int64(1), req.Length())
+				eventList := verifyListNode(t, req, "events", 1)
+				event := verifyListElement(t, eventList, 0)
+				// require.Equal(t, int64(8), event.Length())
```

Comment:
probably want to drop these comments before merge

Reply:
The goal is to get these working, but not being able to control the times the events are emitted makes verifying the properties that are calculated by duration difficult. I was hoping once we were able to control event timing, these could be uncommented and actually tested.